// @clerk/backend/dist/internal.mjs
import {
  AuthStatus,
  TokenType,
  authenticatedMachineObject,
  constants,
  createAuthenticateRequest,
  createBootstrapSignedOutState,
  createClerkRequest,
  createRedirect,
  debugRequestState,
  decorateObjectWithResources,
  getAuthObjectForAcceptedToken,
  getAuthObjectFromJwt,
  invalidTokenAuthObject,
  isMachineToken,
  isMachineTokenByPrefix,
  isTokenTypeAccepted,
  makeAuthObjectSerializable,
  reverificationError,
  reverificationErrorResponse,
  signedInAuthObject,
  signedOutAuthObject,
  stripPrivateDataFromObject,
  unauthenticatedMachineObject,
  verifyMachineAuthToken
} from "./chunk-XZ3TBDTZ.mjs";
import "./chunk-YBVFDYDR.mjs";
import "./chunk-J2CDX2WG.mjs";
import "./chunk-RZ7A7F6X.mjs";
import "./chunk-TOROEX6P.mjs";
export {
  AuthStatus,
  TokenType,
  authenticatedMachineObject,
  constants,
  createAuthenticateRequest,
  createBootstrapSignedOutState,
  createClerkRequest,
  createRedirect,
  debugRequestState,
  decorateObjectWithResources,
  getAuthObjectForAcceptedToken,
  getAuthObjectFromJwt,
  invalidTokenAuthObject,
  isMachineToken,
  isMachineTokenByPrefix,
  isTokenTypeAccepted,
  makeAuthObjectSerializable,
  reverificationError,
  reverificationErrorResponse,
  signedInAuthObject,
  signedOutAuthObject,
  stripPrivateDataFromObject,
  unauthenticatedMachineObject,
  verifyMachineAuthToken
};
